### PR TITLE
implemented debug params, (aka: 100% match?)

### PR DESCRIPTION
**Pull request by danwalkowski**

### CODE REVIEW

File: Branch-SDK/Branch-SDK/Branch.h [hunk 1]
```diff
@@ -431,6 +431,14 @@ typedef NS_ENUM(NSUInteger, BranchPromoCodeUsageType) {
 + (void)setDebug __attribute__((deprecated(("Use the instance method instead"))));
 
 /**
+ Specify additional constant parameters to be included in the response
+ 
+ @param debugParams dictionary of keystrings/valuestrings that will be added to response 
+ */
+- (void) setDeepLinkDebugMode:(NSDictionary *)debugParams;
```
Comment (derrickstaten) [inline Branch-SDK/Branch-SDK/Branch.h]: can you remove the extra space in `(void) setDeepLinkDebugMode`

**Comment by derrickstaten** [comment]
This doesn't appear to be working. I'm not getting back `@{@"DEBUG_PARAM" : @"MONKEY"}` in the deepLinkHandler's params:

```
finished init with params = {
    "+clicked_branch_link" = 0;
    "+is_first_session" = 0;
}
```

**Comment by danwalkowski** [comment]
I'm putting them into the session params and install params, is that the wrong place?

**Comment by danwalkowski** [comment]
this is fixed now.  please thumb @derrickstaten

File: Branch-SDK/Branch-SDK/Branch.m [hunk 2]
```diff
@@ -620,11 +627,25 @@ - (BranchLinkProperties *)getFirstReferringBranchLinkProperties {
 }
 
 - (NSDictionary *)getFirstReferringParams {
-    return [BNCEncodingUtils decodeJsonStringToDictionary:self.preferenceHelper.installParams];
+    NSDictionary *origInstallParams = [BNCEncodingUtils decodeJsonStringToDictionary:self.preferenceHelper.installParams];
```
Comment (derrickstaten) [inline Branch-SDK/Branch-SDK/Branch.m]: let's move this to after the `if` statement. there's no reason to execute this code if `self.deepLinkDebugParams` is true

File: Branch-SDK/Branch-SDK/Branch.m [hunk 3]
```diff
@@ -620,11 +627,25 @@ - (BranchLinkProperties *)getFirstReferringBranchLinkProperties {
...
+        NSMutableDictionary* debugInstallParams = [[BNCEncodingUtils decodeJsonStringToDictionary:self.preferenceHelper.sessionParams] mutableCopy];
+        [debugInstallParams addEntriesFromDictionary:self.deepLinkDebugParams];
+        return debugInstallParams;
+    }
+    return origInstallParams;
 }
 
 - (NSDictionary *)getLatestReferringParams {
-    return [BNCEncodingUtils decodeJsonStringToDictionary:self.preferenceHelper.sessionParams];
+    NSDictionary *origSessionParams = [BNCEncodingUtils decodeJsonStringToDictionary:self.preferenceHelper.sessionParams];
```
Comment (derrickstaten) [inline Branch-SDK/Branch-SDK/Branch.m]: same here -- no need to instantiate this if we aren't going to use it

File: Branch-SDK/Branch-SDK/Branch.m (hunk 2 above)
Comment (danwalkowski) [inline Branch-SDK/Branch-SDK/Branch.m]: That's not correct.  We always need to return the decoded original params.  The if block in the middle just adds more debug params if there are any.

Comment (derrickstaten) [inline Branch-SDK/Branch-SDK/Branch.m]: oh, that's odd. I guess I'm okay with that if @aaustin is okay with it.
I'll test it manually now

On Mon, Jan 4, 2016 at 1:21 PM, Dan Walkowski <notifications@github.com>
wrote:

> In Branch-SDK/Branch-SDK/Branch.m
> <https://github.com/BranchMetrics/iOS-Deferred-Deep-Linking-SDK/pull/263#discussion_r48768876>
> :
>
> > @@ -620,11 +627,25 @@ - (BranchLinkProperties *)getFirstReferringBranchLinkProperties {
> >  }
> >
> >  - (NSDictionary *)getFirstReferringParams {
> > -    return [BNCEncodingUtils decodeJsonStringToDictionary:self.preferenceHelper.installParams];
> > +    NSDictionary *origInstallParams = [BNCEncodingUtils decodeJsonStringToDictionary:self.preferenceHelper.installParams];
>
> That's not correct. We always need to return the decoded original params.
> The if block in the middle just adds more debug params if there are any.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BranchMetrics/iOS-Deferred-Deep-Linking-SDK/pull/263/files#r48768876>
> .
>



-- 
[image: Branch] <https://branch.io/?bmp=xink-sig>
Derrick Staten
Engineer
What is deep linking? <https://t.xink.io/Tracking/Index/Kg4AAGZBAABoTAYA0>
Check out our webinars <https://t.xink.io/Tracking/Index/Kw4AAGZBAABoTAYA0>

**Comment by derrickstaten** [comment]
works for me. Can we do something other than DEBUG_PARAM: MONKEY? How about `foo`:`bar` or something else? then :+1:

**Comment by kylebrowning** [comment]
I noticed a crash around this.

Is this fixed in the most recent version?

```
Fatal Exception: NSInvalidArgumentException
0  CoreFoundation                 0x184762fe0 __exceptionPreprocess
1  libobjc.A.dylib                0x1831c4538 objc_exception_throw
2  CoreFoundation                 0x184769ef4 __methodDescriptionForSelector
3  CoreFoundation                 0x184766f54 ___forwarding___
4  CoreFoundation                 0x184662d4c _CF_forwarding_prep_0
5  League                         0x10044eec8 +[BNCEncodingUtils decodeJsonStringToDictionary:] (BNCEncodingUtils.m:391)
6  League                         0x10045fe30 -[Branch getLatestReferringParams] (Branch.m:810)
7  League                         0x1000ba654 specialized LoginViewController.(autoLogin() -> ()).(closure #1) (LoginViewController.swift:332)
8  League                         0x1000b0400 LoginViewController.autoLogin() -> () (LoginViewController.swift:351)
9  League                         0x1000af4a8 LoginViewController.viewDidAppear(Bool) -> () (LoginViewController.swift:223)
10 League                         0x1000af4f0 @objc LoginViewController.viewDidAppear(Bool) -> () (LoginViewController.swift)
11 UIKit                          0x18a8ac918 -[UIViewController _setViewAppearState:isAnimating:]
12 UIKit                          0x18a965b80 -[UINavigationController viewDidAppear:]
13 UIKit                          0x18a8ac918 -[UIViewController _setViewAppearState:isAnimating:]
14 UIKit                          0x18a91965c __64-[UIViewController viewDidMoveToWindow:shouldAppearOrDisappear:]_block_invoke
15 UIKit                          0x18a9195f8 -[UIViewController _executeAfterAppearanceBlock]
16 UIKit                          0x18ab2169c _runAfterCACommitDeferredBlocks
17 UIKit                          0x18ab13630 _cleanUpAfterCAFlushAndRunDeferredBlocks
18 UIKit                          0x18a88728c _afterCACommitHandler
19 CoreFoundation                 0x1847109a8 __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__
20 CoreFoundation                 0x18470e630 __CFRunLoopDoObservers
21 CoreFoundation                 0x18470ea7c __CFRunLoopRun
22 CoreFoundation                 0x18463eda4 CFRunLoopRunSpecific
23 GraphicsServices               0x1860a9074 GSEventRunModal
24 UIKit                          0x18a8f9c9c UIApplicationMain
25 League                         0x10041b284 main (AppDelegate.swift:22)
26 libdyld.dylib                  0x18364d59c start
```

**Comment by E-B-Smith** [comment]
This is particular issue is fixed in the latest version of the SDK.

What version of the SDK are you using?

**Pull request merged**